docs/extensions/dnsrecord.md

```diff
@@ -0,0 +1,123 @@
+# Contract: `DNSRecord` resources
+
+Every shoot cluster requires external DNS records that are publicly resolvable.
+The management of these DNS records requires provider-specific knowledge which is to be developed outside the Gardener's core repository.
+
+Currently, Gardener uses [`DNSProvider` and `DNSEntry` resources](dns.md). However, this introduces undesired coupling of Gardener to a controller that does not adhere to the Gardener extension contracts. Because of this, we plan to stop using `DNSProvider` and `DNSEntry` resources for Gardener DNS records in the future and use the `DNSRecord` resources described here instead. 
+
+## What does Gardener create DNS records for?
+
+### Internal domain name
+
+Every shoot cluster's kube-apiserver running in the seed is exposed via a load balancer that has a public endpoint (IP or hostname).
+This endpoint is used by end-users and also by system components (that are running in another network, e.g., the kubelet or kube-proxy) to talk to the cluster.
+In order to be robust against changes of this endpoint (e.g., caused due to re-creation of the load balancer or move of the DNS record to another seed cluster) Gardener creates a so-called *internal domain name* for every shoot cluster.
+The *internal domain name* is a publicly resolvable DNS record that points to the load balancer of the kube-apiserver.
+Gardener uses this domain name in the kubeconfigs of all system components, instead of using directly the load balancer endpoint.
+This way Gardener does not need to recreate all kubeconfigs if the endpoint changes - it just needs to update the DNS record.
+
+### External domain name
+
+The internal domain name is not configurable by end-users directly but configured by the Gardener administrator.
+However, end-users usually prefer to have another DNS name, maybe even using their own domain sometimes to access their Kubernetes clusters.
+Gardener supports that by creating another DNS record, named *external domain name*, that actually points to the *internal domain name*.
+The kubeconfig handed out to end-users does contain this *external domain name*, i.e., users can access their clusters with the DNS name they like to.
+
+As not every end-user has an own domain it is possible for Gardener administrators to configure so-called *default domains*.
+If configured, shoots that do not specify a domain explicitly get an *external domain name* based on a default domain (unless explicitly stated that this shoot should not get an external domain name (`.spec.dns.provider=unmanaged`).
+
+### Ingress domain name (deprecated)
+
+Gardener allows to deploy a `nginx-ingress-controller` into a shoot cluster (deprecated).
+This controller is exposed via a public load balancer (again, either IP or hostname).
+Gardener creates a wildcard DNS record pointing to this load balancer.
+`Ingress` resources can later use this wildcard DNS record to expose underlying applications.
```

Comment:
Even though nginx addon is deprecated, I think the domain is not, i.e. shoot owner can deploy their own ingress controllers and ask gardener to create the dns record. Probably, this is done now via the shoot-dns extension when the LB service is annotated and not handled by gardenlet. Unfortunately, I have no resource that I can refer.

Reply:
This was part of #4231